File: khalti-android/src/main/java/com/khalti/checkOut/ebanking/EBankingContract.java
package com.khalti.checkOut.ebanking;

import com.khalti.checkOut.ebanking.helper.BankPojo;
import com.khalti.checkOut.ebanking.helper.BankingData;

import java.util.HashMap;
import java.util.List;

import rx.Observable;

public interface EBankingContract {
    interface View {

        void toggleIndented(boolean show);

        void setUpList(List<BankPojo> bankList);

        void showIndentedNetworkError();

        void showIndentedError(String error);

        void openMobileForm(BankingData bankingData);

        Observable<HashMap<String, String>> getItemClickObservable();

        HashMap<String, Observable<Void>> setOnClickListener();


        Observable<CharSequence> setSearchListener();

        Observable<Integer> filterList(String text);

        void toggleSearch(boolean show);

        void toggleSearchError(boolean show);

        boolean hasNetwork();

        void setPresenter(Presenter presenter);
    }

    interface Presenter {

        void onCreate();

        void onDestroy();
    }

    interface Model {

        Observable<List<BankPojo>> fetchBankList();

        void unSubscribe();
    }
}

File: khalti-android/src/main/java/com/khalti/checkOut/ebanking/EBankingContract.kt
package com.khalti.checkOut.ebanking

import com.khalti.base.LifeCycle
import com.khalti.checkOut.api.Result
import com.khalti.checkOut.ebanking.helper.BankPojo
import com.khalti.checkOut.ebanking.helper.BankingData
import com.khalti.signal.Signal

import java.util.HashMap

import rx.Observable

interface EBankingContract {
    interface View {

        fun toggleIndented(show: Boolean)

        fun toggleSearch(show: Boolean)

        fun toggleSearchError(show: Boolean)

        fun setupList(bankList: MutableList<BankPojo>): Signal<Map<String, String>>

        fun showIndentedNetworkError()

        fun showIndentedError(error: String)

        fun openMobileForm(bankingData: BankingData)

        fun setOnClickListener(): Map<String, Signal<Any>>

        fun setSearchListener(): Signal<String>

        fun filterList(text: String): Signal<Int>

        fun hasNetwork(): Boolean

        fun setPresenter(presenter: Presenter)
    }

    interface Presenter : LifeCycle

    interface Model {

        suspend fun fetchBankList(): Result<List<BankPojo>>
    }
}
